apps/prairielearn/src/sync/syncFromDisk.ts

```diff
@@ -47,9 +76,20 @@ export async function syncDiskToSqlWithLock(
 
   logger.info('Loading info.json files from course repository');
 
-  const courseData = await timed('Loaded course data from disk', () =>
-    courseDB.loadFullCourse(courseId, courseDir),
+  const { courseData, sharingSyncError } = await timed('Loaded course data from disk', () =>
+    validateCourseData(courseId, courseDir, logger),
   );
+  if (sharingSyncError) {
+    const courseDataHasErrors = courseDB.courseDataHasErrors(courseData);
+    const courseDataHasErrorsOrWarnings = courseDB.courseDataHasErrorsOrWarnings(courseData);
+    return {
+      sharingSyncError: true,
+      hadJsonErrors: courseDataHasErrors,
+      hadJsonErrorsOrWarnings: courseDataHasErrorsOrWarnings,
+      courseId,
+      courseData,
+    };
```

Comment:
Nit/bikeshed: I'm thinking a bit about the notion of ["make invalid states unrepresentable"](https://geeklaunch.io/blog/make-invalid-states-unrepresentable/). It's not exactly the same concept, but: at this point, we don't actually _know for sure_ if there are JSON errors or warnings because the below syncing operations can actually add more warnings/errors. Therefore, it would be invalid for a caller to try to check either of those if a sharing sync error was present. In fact, it should also be invalid for them to loo

What if we represent the possible results of `syncDiskToSqlWithLock` as a union of types to force the caller to handle all the possibilities here? Something like this:

```ts
interface SyncResultSharingError {
  status: 'sharing_error';
  courseId: string;
}

interface SyncResultSuccess {
  status: 'success';
  hadJsonErrors: boolean;
  hadJsonErrorsOrWarnings: boolean;
  courseId: string;
  courseData: CourseData;
}

type SyncResult = SyncResultSharingError | SyncResultSuccess;
```

I'm going through all this trouble because syncing+sharing is complicated and will only get moreso, and I want to encode as much as possible in the type system so that it's harder for us to mess things up in the future.

Reply:
Self-nit: I don't like "success" as a status here, since technically some forms of sync failure fall under that. Maybe "complete" is better?

Reply:
Love this idea, all done! I did go with `SyncResultsComplete` rather than using 'success'